ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp

```diff
@@ -0,0 +1,405 @@
+//*****************************************************************************
+// Copyright 2017-2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <functional>
+#include <onnx/onnx_pb.h>
+#include <stack>
+
+#include "ngraph/check.hpp"
+#include "onnx_import/editor/detail/subgraph_extraction.hpp"
+
+using namespace ngraph::onnx_import;
+
+namespace
+{
+    void validate_node_index(const ONNX_NAMESPACE::GraphProto& graph, const int node_idx)
+    {
+        NGRAPH_CHECK(
+            node_idx >= 0 && node_idx < graph.node_size(),
+            "The specified node index is out of range of nodes in the original model(idx: ",
+            std::to_string(node_idx),
+            "; nodes count in the model: ",
+            std::to_string(graph.node_size()),
+            ")");
+    }
+
+    struct CheckIfHasName
+    {
+    };
+
+    template <typename T>
+    std::function<bool(const T&)> name_equals(const std::string& name)
+    {
+        return [&name](const T& onnx_object) -> bool { return onnx_object.name() == name; };
+    }
+
+    template <typename T>
+    std::function<bool(const T&)> name_equals(const std::string& name, CheckIfHasName)
+    {
+        return [&name](const T& onnx_object) -> bool {
+            return onnx_object.has_name() && onnx_object.name() == name;
+        };
+    }
+
+    std::function<bool(const std::string&)> is_equal_to(const std::string& other)
+    {
+        return [&other](const std::string& s) { return s == other; };
+    }
+
+    template <typename Container>
+    bool already_exists(const Container& items, const std::string& name)
+    {
+        return std::any_of(
+            items.begin(), items.end(), name_equals<typename Container::value_type>(name));
```

Comment:
Can't use ADL because protobuf does not provide the std::begin nor std::end overload for their `RepeatedPtrField` class. I'll switch to free functions instead of method calls though.

Reply:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c165-use-using-for-customization-points